aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+  int64_t right_ind = length - 1;
+  int64_t mid_ind;
+  bool done_searching = false;
+
+  while (!done_searching) {
+    mid_ind = (left_ind+right_ind) >> 1;
+    scalar_t mid_val = sorted_arr[mid_ind*sorted_arr_stride];
+
+    if (mid_val > search_val) {
+      right_ind = mid_ind-1;
+    } else if((mid_val == search_val) && (
+      (mid_ind == length - 1) || (sorted_arr[(mid_ind+1)*sorted_arr_stride] != search_val)
+    )) {
+      done_searching = true;
+      *found = true;
+    } else {
+      left_ind = mid_ind+1;
+    }
+
+    if (left_ind > right_ind) {
+      done_searching = true;
+      *found = false;
+      mid_ind = -1;
+    }
+  }
+
+  return mid_ind;
+}
+
+Tensor& bmm_out_sparse_cpu(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
+  TORCH_CHECK(self.sparse_dim() == 3, "bmm_sparse: Tensor 'self' must have 3 sparse dims, but has ", self.sparse_dim());
+  TORCH_CHECK(mat2.dim() == 3, "bmm_sparse: Tensor 'mat2' must have 3 dims, but has ", mat2.dim());
+
+  TORCH_CHECK(self.size(0) == mat2.size(0), "bmm_sparse: 'self.size(0)' and 'mat2.size(0)' must match");
+  TORCH_CHECK(self.size(2) == mat2.size(1), "bmm_sparse: 'self.size(2)' and 'mat2.size(1)' must match");
+
+  result.resize_({self.size(0), self.size(1), mat2.size(2)});
+
+  if (self._nnz() == 0) {
+    result.zero_();
+    return result;
+  }
+
+  // First need to coalesce to get all of the first dimension indices
+  // in order since we'll be sending each matrix into the MM operation
+  SparseTensor self_coalesced;
+
+  if (!self.is_coalesced()){
+    self_coalesced = self.coalesce();
+  } else {
+    self_coalesced = self;
+  }
+
+  int64_t nnz =        self_coalesced._nnz();
+  LongTensor indices = self_coalesced._indices();
+  Tensor values =      self_coalesced._values();
+
+  LongTensor indices_dim0 = indices[0];
+  int64_t indices_dim0_stride = indices_dim0.stride(0);
+  int64_t* indices_dim0_ptr = indices_dim0.data_ptr<int64_t>();
+  LongTensor indices_dim1_dim2 = indices.slice(0, 1, 3);
+
+  int64_t dim_i = self_coalesced.size(1);
+  int64_t dim_j = self_coalesced.size(2);
+  int64_t dim_k = mat2.size(2);
+
+  Scalar beta = 0;
+  Tensor t = at::ones({self_coalesced.size(1)});
+  Scalar alpha = 1;
+
+  int64_t mat_el_begin_idx = 0;
+
+  int64_t num_matrices = self_coalesced.size(0);
+
+  // Iterate through each set of 2D matrices within the 3D
+  // tensor inputs, performing a matrix multiply with each one.
+  int64_t start_mat_num = indices_dim0_ptr[0];
+  for (int64_t cur_mat_num = 0;
+    (cur_mat_num < num_matrices);
+    cur_mat_num++
+  ) {
+    // If there are sparse matrices at the beginning or end that
+    // have all zero elements, we need to zero out the result matrix.
+    if ((cur_mat_num < start_mat_num) || (mat_el_begin_idx >= nnz)) {
```

Comment:
So, if I understand correctly, if there's a big stack of "zero" matrices at the end, we're just going to loop through them one by one and repeatedly call `zero_`. I wonder if it wouldn't be smarter to just up front zero the result tensor, and then walk down the index splits (which now itself can be sparse). Well, it's probably not worth it as I doubt people are going to do batched sparse matmul with a bunch of zero matrices, that's dumb lol.

Reply:
Right, no one should really be using so many zero matrices, but this function should still support that. In most cases it will be more performant to avoid always zeroing out the entire thing up front.